src/backend/distributed/master/master_repair_shards.c

```diff
@@ -237,45 +332,314 @@ RecreateTableDDLCommandList(Oid relationId)
+
+	sourcePlacement = SearchShardPlacementInList(shardPlacementList, sourceNodeName,
+												 sourceNodePort, missingSourceOk);
+	if (sourcePlacement->shardState != FILE_FINALIZED)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("source placement must be in finalized state")));
+	}
+
+	targetPlacement = SearchShardPlacementInList(shardPlacementList, targetNodeName,
+												 targetNodePort, missingTargetOk);
+	if (targetPlacement->shardState != FILE_INACTIVE)
 	{
-		copySuccessful = true;
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("target placement must be in inactive state")));
 	}
+}
+
+
+/*
+ * EnsureShardCanBeCopied checks if the given shard has a healthy placement in the source.
+ */
+static void
+EnsureShardCanBeCopied(int64 shardId, char *sourceNodeName, int32 sourceNodePort,
+					   char *targetNodeName, int32 targetNodePort)
+{
+	List *shardPlacementList = ShardPlacementList(shardId);
+	ShardPlacement *sourcePlacement = NULL;
+	ShardPlacement *targetPlacement = NULL;
+	bool missingSourceOk = false;
+	bool missingTargetOk = true;
+
+	sourcePlacement = SearchShardPlacementInList(shardPlacementList, sourceNodeName,
+												 sourceNodePort, missingSourceOk);
+	if (sourcePlacement->shardState != FILE_FINALIZED)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("source placement must be in finalized state")));
+	}
+
+	targetPlacement = SearchShardPlacementInList(shardPlacementList, targetNodeName,
+												 targetNodePort, missingTargetOk);
+	if (targetPlacement != NULL)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("shard %ld already exist in target placement", shardId)));
+	}
+}
+
+
+/*
+ * EnsureShardCanBeMoved checks if the given shard has a placement in the source node but
+ * not on the target node. It is important to note that SearchShardPlacementInList
+ * function already generates error if given shard does not have a placement in the
+ * source node. Therefore we do not perform extra check.
+ */
+static void
+EnsureShardCanBeMoved(int64 shardId, char *sourceNodeName, int32 sourceNodePort)
+{
+	List *shardPlacementList = ShardPlacementList(shardId);
+	bool missingSourceOk = false;
 
-	return copySuccessful;
+	/* Actual check is done in SearchShardPlacementInList  */
+	SearchShardPlacementInList(shardPlacementList, sourceNodeName, sourceNodePort,
+							   missingSourceOk);
+}
+
+
+/*
+ * CopyShardPlacement copies a shard along with its co-located shards from a source node
+ * to target node. CopyShardPlacement does not make any checks about state of the shards.
+ * It is caller's responsibility to make those checks if they are necessary. If
+ * dropTargetPlacement is set to true, this function will remove existing placement row
+ * then re-creates it with FILE_FINALIZED state.
+ */
+static void
+CopyShardPlacement(int64 shardId, char *sourceNodeName, int32 sourceNodePort,
```

Comment:
Ok, I found what we should do.

```C
RepairShardPlacement()
   DeleteShardPlacementRow()
   CopyShardPlacement()

def master_copy_shard_placement:
   if repair 
     EnsureCanBeRepaird
     RepairShardPlacement
  else
    EnsureCanBeCopied
    CopyShrdPlacement
```


Then, remove the `deleteTargetPlacementRow` parameter from this function. Because, I think it is not the right way to pass `deleteTargetPlacementRow` to this function at all.